src/Compression/CompressionCodecDeflate.cpp

```diff
@@ -0,0 +1,357 @@
+    {
+        DeflateJobHWPool::instance().releaseJob(job_id);
+        return job_ptr->total_out;
+    }
+    else
+    {
+        LOG_WARNING(log, "HardwareCodecDeflate::doDecompressData fail ->status: '{}' ", static_cast<size_t>(status));
+        DeflateJobHWPool::instance().releaseJob(job_id);
+        return 0;
+    }
+}
+
+uint32_t HardwareCodecDeflate::doDecompressDataReq(const char * source, uint32_t source_size, char * dest, uint32_t uncompressed_size)
+{
+    uint32_t job_id = 0;
+    qpl_job * job_ptr = DeflateJobHWPool::instance().acquireJob(&job_id);
+    if (job_ptr == nullptr)
+    {
+        LOG_WARNING(log, "HardwareCodecDeflate::doDecompressDataReq acquireJob fail!");
+        return 0;
+    }
+    qpl_status status;
+
+    // Performing a decompression operation
+    job_ptr->op = qpl_op_decompress;
+    job_ptr->next_in_ptr = reinterpret_cast<uint8_t *>(const_cast<char *>(source));
+    job_ptr->next_out_ptr = reinterpret_cast<uint8_t *>(dest);
+    job_ptr->available_in = source_size;
+    job_ptr->available_out = uncompressed_size;
+    job_ptr->flags = QPL_FLAG_FIRST | QPL_FLAG_LAST;
+
+    // Decompression
+    status = qpl_submit_job(job_ptr);
+    if (QPL_STS_OK == status)
+    {
+        jobDecompAsyncMap.insert(std::make_pair(job_id, job_ptr));
+        return job_id;
+    }
+    else
+    {
+        DeflateJobHWPool::instance().releaseJob(job_id);
+        LOG_WARNING(log, "HardwareCodecDeflate::doDecompressDataReq fail ->status: '{}' ", static_cast<size_t>(status));
+        return 0;
+    }
+}
+
+void HardwareCodecDeflate::doDecompressDataFlush()
+{
+    uint32_t job_id = 0;
+    qpl_job * job_ptr = nullptr;
+
+
+    std::map<uint32_t, qpl_job *>::iterator it;
+    uint32_t n_jobs_processing = jobDecompAsyncMap.size();
+    it = jobDecompAsyncMap.begin();
+
+    while (n_jobs_processing)
+    {
+        job_id = it->first;
+        job_ptr = it->second;
+
+        if (QPL_STS_BEING_PROCESSED == qpl_check_job(job_ptr))
+        {
+            it++;
+        }
+        else
+        {
+            DeflateJobHWPool::instance().releaseJob(job_id);
+            it = jobDecompAsyncMap.erase(it);
+            n_jobs_processing--;
+            if (n_jobs_processing <= 0)
+                break;
+        }
+        if (it == jobDecompAsyncMap.end())
+        {
+            it = jobDecompAsyncMap.begin();
+            _tpause(1, __rdtsc() + 1000);
+        }
+    }
+}
+//SoftwareCodecDeflate
+SoftwareCodecDeflate::SoftwareCodecDeflate()
+{
+    jobSWPtr = nullptr;
+}
+
+SoftwareCodecDeflate::~SoftwareCodecDeflate()
+{
+    if (nullptr != jobSWPtr)
+        qpl_fini_job(jobSWPtr);
+}
+
+qpl_job * SoftwareCodecDeflate::getJobCodecPtr()
+{
+    if (nullptr == jobSWPtr)
+    {
+        qpl_status status;
+        uint32_t size = 0;
+        // Job initialization
+        status = qpl_get_job_size(qpl_path_software, &size);
+        if (status != QPL_STS_OK)
+        {
+            throw Exception(
```

Comment:
ClickHouse provides also modern fmt-based exceptions:

```
throw Exception(
    ErrorCodes::CANNOT_COMPRESS,
    "SoftwareCodecDeflate::getJobCodecPtr -> qpl_get_job_size fail: {}", status)
```

Reply:
User Joe Average will despair when an exception with such text is returned. We need something more descriptive, e.g. "Initialization of DEFLATE codec failed (details: qpl_get_job_size_failed with error code {} - XXX" where "XXX" is ideally a hint about a next step, e.g. XXX = "is AVX2 enabled?" etc.

Reply:
Done